File: src/config/loader.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { parse } from "smol-toml";
import { debug } from "../debug.ts";
import { getHeddleHome } from "./paths.ts";

export type ApprovalMode = "suggest" | "auto-edit" | "full-auto" | "plan" | "yolo";

const VALID_APPROVAL_MODES: ReadonlySet<string> = new Set(["suggest", "auto-edit", "full-auto", "plan", "yolo"]);

export interface HeddleConfig {
	// ── Identity ──
	apiKey?: string;

	// ── Provider/API ──
	model: string;
	weakModel?: string;
	editorModel?: string;
	maxTokens?: number;
	temperature?: number;
	baseUrl?: string;

	// ── Session ──
	systemPrompt?: string;
	approvalMode?: ApprovalMode;
	instructions?: string[];
	tools?: string[];
	doomLoopThreshold?: number;
	budgetLimit?: number;
}

/** Type aliases for consumer clarity. */
export type ProviderFields = Pick<
	HeddleConfig,
	"model" | "weakModel" | "editorModel" | "maxTokens" | "temperature" | "baseUrl"
>;
export type SessionFields = Pick<
	HeddleConfig,
	"systemPrompt" | "approvalMode" | "instructions" | "tools" | "doomLoopThreshold" | "budgetLimit"
>;

const DEFAULTS: HeddleConfig = {
	model: "openrouter/free",
};

/** Parse a TOML file into a plain object. Returns {} on missing/malformed files. */
function loadToml(path: string): Record<string, unknown> {
	if (!existsSync(path)) return {};
	try {
		const content = readFileSync(path, "utf-8");
		if (!content.trim()) return {};
		return parse(content) as Record<string, unknown>;
	} catch {
		return {};
	}
}

/** Map raw TOML keys to HeddleConfig fields. */
function toConfig(raw: Record<string, unknown>): Partial<HeddleConfig> {
	const config: Partial<HeddleConfig> = {};

	// String fields
	if (typeof raw.model === "string") config.model = raw.model;
	if (typeof raw.api_key === "string") config.apiKey = raw.api_key;
	if (typeof raw.system_prompt === "string") config.systemPrompt = raw.system_prompt;
	if (typeof raw.weak_model === "string") config.weakModel = raw.weak_model;
	if (typeof raw.editor_model === "string") config.editorModel = raw.editor_model;
	if (typeof raw.base_url === "string") config.baseUrl = raw.base_url;

	// Number fields
	if (typeof raw.max_tokens === "number") config.maxTokens = raw.max_tokens;
	if (typeof raw.temperature === "number") config.temperature = raw.temperature;
	if (typeof raw.doom_loop_threshold === "number") config.doomLoopThreshold = raw.doom_loop_threshold;
	if (typeof raw.budget_limit === "number") config.budgetLimit = raw.budget_limit;

	// Approval mode — validate against allowed values
	if (typeof raw.approval_mode === "string" && VALID_APPROVAL_MODES.has(raw.approval_mode)) {
		config.approvalMode = raw.approval_mode as ApprovalMode;
	}

	// Instructions — must be an array of strings
	if (Array.isArray(raw.instructions)) {
		const filtered = raw.instructions.filter((item): item is string => typeof item === "string");
		if (filtered.length > 0) config.instructions = filtered;
	}

	// Tools — must be an array of strings
	if (Array.isArray(raw.tools)) {
		const filtered = raw.tools.filter((item): item is string => typeof item === "string");
		if (filtered.length > 0) config.tools = filtered;
	}

	return config;
}

/**
 * Load config with layered merging: defaults → global → local → env vars.
 * @param localDir Path to the local .heddle directory (defaults to cwd/.heddle)
 */
export function loadConfig(localDir?: string): HeddleConfig {
	const globalPath = join(getHeddleHome(), "config.toml");
	const localPath = localDir ? join(localDir, "config.toml") : undefined;

	const globalRaw = loadToml(globalPath);
	const localRaw = localPath ? loadToml(localPath) : {};

	const merged: HeddleConfig = {
		...DEFAULTS,
		...toConfig(globalRaw),
		...toConfig(localRaw),
	};

	// Env vars override everything
	if (process.env.HEDDLE_MODEL) merged.model = process.env.HEDDLE_MODEL;
	if (process.env.OPENROUTER_API_KEY) merged.apiKey = process.env.OPENROUTER_API_KEY;
	if (process.env.HEDDLE_BASE_URL) merged.baseUrl = process.env.HEDDLE_BASE_URL;

	// Numeric env vars with isNaN guard
	const maxTokensEnv = process.env.HEDDLE_MAX_TOKENS;
	if (maxTokensEnv && !Number.isNaN(Number(maxTokensEnv))) {
		merged.maxTokens = Number(maxTokensEnv);
	}

	const temperatureEnv = process.env.HEDDLE_TEMPERATURE;
	if (temperatureEnv && !Number.isNaN(Number(temperatureEnv))) {
		merged.temperature = Number(temperatureEnv);
	}

	if (process.env.HEDDLE_WEAK_MODEL) merged.weakModel = process.env.HEDDLE_WEAK_MODEL;

	const approvalModeEnv = process.env.HEDDLE_APPROVAL_MODE;
	if (approvalModeEnv && VALID_APPROVAL_MODES.has(approvalModeEnv)) {
		merged.approvalMode = approvalModeEnv as ApprovalMode;
	}

	const toolsEnv = process.env.HEDDLE_TOOLS;
	if (toolsEnv) {
		const parsed = toolsEnv
			.split(",")
			.map((t) => t.trim())
			.filter(Boolean);
		if (parsed.length > 0) merged.tools = parsed;
	}

	debug("config", "loaded:", merged);
	return merged;
}

File: src/config/loader.rs
//! Layered TOML config loader (defaults → global → local → env).
//! Mirrors `ts-src/config/loader.ts`.

use std::path::Path;

use serde::{Deserialize, Serialize};
use toml::Value as TomlValue;

use crate::config::features::FeatureFlagsOverride;
use crate::config::paths::{get_heddle_home, get_local_heddle_dir};
use crate::debug::debug;
use crate::hooks::loader::load_hooks;
use crate::hooks::types::ResolvedHooksConfig;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ApprovalMode {
    Suggest,
    AutoEdit,
    FullAuto,
    Plan,
    Yolo,
}

impl ApprovalMode {
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "suggest" => Some(Self::Suggest),
            "auto-edit" => Some(Self::AutoEdit),
            "full-auto" => Some(Self::FullAuto),
            "plan" => Some(Self::Plan),
            "yolo" => Some(Self::Yolo),
            _ => None,
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Suggest => "suggest",
            Self::AutoEdit => "auto-edit",
            Self::FullAuto => "full-auto",
            Self::Plan => "plan",
            Self::Yolo => "yolo",
        }
    }
}

#[derive(Debug, Clone, Default)]
pub struct PermissionsLayer {
    pub allow: Vec<String>,
    pub deny: Vec<String>,
    pub ask: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct HeddleConfig {
    pub api_key: Option<String>,

    pub model: String,
    pub weak_model: Option<String>,
    pub editor_model: Option<String>,
    pub max_tokens: Option<u64>,
    pub temperature: Option<f64>,
    pub base_url: Option<String>,

    pub system_prompt: Option<String>,
    pub approval_mode: Option<ApprovalMode>,
    pub instructions: Option<Vec<String>>,
    pub tools: Option<Vec<String>>,
    pub doom_loop_threshold: Option<u32>,
    pub budget_limit: Option<f64>,
    pub compact_trigger: Option<f64>,
    pub prune_protect: Option<u64>,
    pub prune_minimum: Option<u32>,
    pub compact_buffer: Option<f64>,
    pub features: Option<FeatureFlagsOverride>,

    pub permissions_layers: Option<Vec<PermissionsLayer>>,
    pub hooks: Option<ResolvedHooksConfig>,
}

impl Default for HeddleConfig {
    fn default() -> Self {
        Self {
            api_key: None,
            model: "openrouter/free".to_string(),
            weak_model: None,
            editor_model: None,
            max_tokens: None,
            temperature: None,
            base_url: None,
            system_prompt: None,
            approval_mode: None,
            instructions: None,
            tools: None,
            doom_loop_threshold: None,
            budget_limit: None,
            compact_trigger: None,
            prune_protect: None,
            prune_minimum: None,
            compact_buffer: None,
            features: None,
            permissions_layers: None,
            hooks: None,
        }
    }
}

fn load_toml(path: &Path) -> TomlValue {
    if !path.exists() {
        return TomlValue::Table(Default::default());
    }
    let content = match std::fs::read_to_string(path) {
        Ok(c) => c,
        Err(_) => return TomlValue::Table(Default::default()),
    };
    if content.trim().is_empty() {
        return TomlValue::Table(Default::default());
    }
    content
        .parse::<TomlValue>()
        .unwrap_or_else(|_| TomlValue::Table(Default::default()))
}

fn as_str(v: &TomlValue) -> Option<&str> {
    v.as_str()
}

fn as_int(v: &TomlValue) -> Option<i64> {
    v.as_integer()
}

fn as_float(v: &TomlValue) -> Option<f64> {
    v.as_float().or_else(|| v.as_integer().map(|i| i as f64))
}

fn as_bool(v: &TomlValue) -> Option<bool> {
    v.as_bool()
}

fn as_string_array(v: &TomlValue) -> Option<Vec<String>> {
    let arr = v.as_array()?;
    let filtered: Vec<String> = arr
        .iter()
        .filter_map(|x| x.as_str().map(|s| s.to_string()))
        .collect();
    if filtered.is_empty() {
        None
    } else {
        Some(filtered)
    }
}

fn apply_raw(config: &mut HeddleConfig, raw: &TomlValue) {
    let table = match raw.as_table() {
        Some(t) => t,
        None => return,
    };

    if let Some(s) = table.get("model").and_then(as_str) {
        config.model = s.to_string();
    }
    if let Some(s) = table.get("api_key").and_then(as_str) {
        config.api_key = Some(s.to_string());
    }
    if let Some(s) = table.get("system_prompt").and_then(as_str) {
        config.system_prompt = Some(s.to_string());
    }
    if let Some(s) = table.get("weak_model").and_then(as_str) {
        config.weak_model = Some(s.to_string());
    }
    if let Some(s) = table.get("editor_model").and_then(as_str) {
        config.editor_model = Some(s.to_string());
    }
    if let Some(s) = table.get("base_url").and_then(as_str) {
        config.base_url = Some(s.to_string());
    }

    if let Some(n) = table.get("max_tokens").and_then(as_int) {
        config.max_tokens = Some(n as u64);
    }
    if let Some(n) = table.get("temperature").and_then(as_float) {
        config.temperature = Some(n);
    }
    if let Some(n) = table.get("doom_loop_threshold").and_then(as_int) {
        config.doom_loop_threshold = Some(n as u32);
    }
    if let Some(n) = table.get("budget_limit").and_then(as_float) {
        config.budget_limit = Some(n);
    }
    if let Some(n) = table.get("compact_trigger").and_then(as_float) {
        config.compact_trigger = Some(n);
    }
    if let Some(n) = table.get("prune_protect").and_then(as_int) {
        config.prune_protect = Some(n as u64);
    }
    if let Some(n) = table.get("prune_minimum").and_then(as_int) {
        config.prune_minimum = Some(n as u32);
    }
    if let Some(n) = table.get("compact_buffer").and_then(as_float) {
        config.compact_buffer = Some(n);
    }

    if let Some(am) = table
        .get("approval_mode")
        .and_then(as_str)
        .and_then(ApprovalMode::from_str)
    {
        config.approval_mode = Some(am);
    }

    if let Some(arr) = table.get("instructions").and_then(as_string_array) {
        config.instructions = Some(arr);
    }
    if let Some(arr) = table.get("tools").and_then(as_string_array) {
        config.tools = Some(arr);
    }

    if let Some(feat) = table.get("features").and_then(|v| v.as_table()) {
        let mut over = FeatureFlagsOverride::default();
        if let Some(b) = feat.get("history").and_then(as_bool) {
            over.history = Some(b);
        }
        if let Some(b) = feat.get("usage_data").and_then(as_bool) {
            over.usage_data = Some(b);
        }
        if let Some(b) = feat.get("facets").and_then(as_bool) {
            over.facets = Some(b);
        }
        if let Some(b) = feat.get("file_history").and_then(as_bool) {
            over.file_history = Some(b);
        }
        if let Some(b) = feat.get("paste_cache").and_then(as_bool) {
            over.paste_cache = Some(b);
        }
        if let Some(b) = feat.get("status_line").and_then(as_bool) {
            over.status_line = Some(b);
        }
        if let Some(b) = feat.get("hooks").and_then(as_bool) {
            over.hooks = Some(b);
        }
        if let Some(b) = feat.get("tasks").and_then(as_bool) {
            over.tasks = Some(b);
        }
        config.features = Some(over);
    }
}

fn extract_permissions(raw: &TomlValue) -> Option<PermissionsLayer> {
    let perms = raw.as_table()?.get("permissions")?.as_table()?;
    let to_strings = |key: &str| -> Vec<String> {
        perms
            .get(key)
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|x| x.as_str().map(|s| s.to_string()))
                    .collect()
            })
            .unwrap_or_default()
    };
    let allow = to_strings("allow");
    let deny = to_strings("deny");
    let ask = to_strings("ask");
    if allow.is_empty() && deny.is_empty() && ask.is_empty() {
        None
    } else {
        Some(PermissionsLayer { allow, deny, ask })
    }
}

/// Load config from defaults → global → local → env vars.
pub fn load_config(local_dir: Option<&Path>) -> HeddleConfig {
    let global_path = get_heddle_home().join("config.toml");
    let local_root = local_dir
        .map(Path::to_path_buf)
        .unwrap_or_else(get_local_heddle_dir);
    let local_path = local_root.join("config.toml");

    let global_raw = load_toml(&global_path);
    let local_raw = load_toml(&local_path);

    let mut merged = HeddleConfig::default();
    apply_raw(&mut merged, &global_raw);
    apply_raw(&mut merged, &local_raw);

    let mut layers: Vec<PermissionsLayer> = Vec::new();
    if let Some(l) = extract_permissions(&global_raw) {
        layers.push(l);
    }
    if let Some(l) = extract_permissions(&local_raw) {
        layers.push(l);
    }
    if !layers.is_empty() {
        merged.permissions_layers = Some(layers);
    }

    let hooks = load_hooks(&global_raw, &local_raw);
    if !hooks.is_empty() {
        merged.hooks = Some(hooks);
    }

    if let Ok(v) = std::env::var("HEDDLE_MODEL") {
        merged.model = v;
    }
    if let Ok(v) = std::env::var("OPENROUTER_API_KEY") {
        merged.api_key = Some(v);
    }
    if let Ok(v) = std::env::var("HEDDLE_BASE_URL") {
        merged.base_url = Some(v);
    }
    if let Ok(v) = std::env::var("HEDDLE_MAX_TOKENS") {
        if let Ok(n) = v.parse::<u64>() {
            merged.max_tokens = Some(n);
        }
    }
    if let Ok(v) = std::env::var("HEDDLE_TEMPERATURE") {
        if let Ok(n) = v.parse::<f64>() {
            merged.temperature = Some(n);
        }
    }
    if let Ok(v) = std::env::var("HEDDLE_WEAK_MODEL") {
        merged.weak_model = Some(v);
    }
    if let Ok(v) = std::env::var("HEDDLE_APPROVAL_MODE") {
        if let Some(am) = ApprovalMode::from_str(&v) {
            merged.approval_mode = Some(am);
        }
    }
    if let Ok(v) = std::env::var("HEDDLE_TOOLS") {
        let parsed: Vec<String> = v
            .split(',')
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .collect();
        if !parsed.is_empty() {
            merged.tools = Some(parsed);
        }
    }

    debug("config", "loaded (api_key REDACTED if present)");
    merged
}
